### ISSUE TITLE
Copy link to Paragraph can copy from the wrong instance of OneNote (as does XML view)

### ISSUE THREAD

**Issue by jasonjac2** [stevencohn/OneMore#1563]
# Problem to Solve
I have 8 OneNote windows open. I use OM copy link to paragraph from a window. If there is another windows that has the *same page open* (which is often the case), the copy Link to Paragraph *MAY* copy the one from the wrong windows (not the one I am on).  

The impact is the link has the wrong text and takes you to the wrong part of the page.  This can be very confusing.

It does the same thing with the "XML viewer", i.e. the XML will show the selection as the one from the wrong window.

I haven't done any additional testing, but my assumption is, it is either using the first instance that has the correct page open (instead of the one with focus).

I think I noticed it a couple of versions ago, but only just realised when it happens as it take a certain set of circumstanced.

I have tried on 2 different PCs with different pages.  You can show the problem with just 2 instances of OneNote.

Steps:
- Open Onenote
- Goto a page with headings (longish)
- Click into a para
- Open a second instance of onenote (ctrl+m) on the same page
- Goto a different part of the page (so that you would have to scroll to get to the 1st part).
- click into a para
- On the second instance, "copy link to para"
- Paste this somewhere (word / the page you were on in the second instancet)
- Note the text of the link will show the para from the 1st instance
- Click the link
- It will take you to the part of the page the 1st instances cursor was.

# Environment (if applicable)
Version 6.5.1 with OneNote 16.0.17928.20114

# Additional Context

THE FINE PRINT This is a hobby project that I started for myself. I'm happy to take requests and will promise to
consider each one carefully but offer no guarantees that I will ultimately agree to their
usefulness or have time to implement any of them. If OneNote offers a "close enough" work-around
then I will probably reject the request - you've been warned!

**Comment by jasonjac2** [comment]
I have added steps to the 1st post to make it clear how to reproduce.

**Comment by jasonjac2** [comment]
I was just doing system documentation so had 2 virtual desktop (i.e. one W11 PC with 2 virtual desktops).  I was on the 2nd desktop and was copying links to a para and then pasting into another instance.  Because I was on the same page on the main virtual desktop, it kept copying the same link regardless of where I was on the page :-).

**Comment by jasonjac2** [comment]
Have you managed to reproduce this / have you had time to look?

**Comment by stevencohn** [comment]
TBH I haven't tried. But TBH, I believe you and can envision exactly why it would happen. Solutions might include:

1. Take advantage of a known ID on each page (the omID I've introduced in other areas) and rely on that for a unique page identifier instead of the OneNote pageId.
2. Find a simpler way to identify the same page across instances or even machines
3. Restrict command to single window instances; this is not the most user friendly

**Comment by jasonjac2** [comment]
From your reply, I'm not sure I have described it quite right - is the issue clear?  i.e. it is like when your command asks ON for the current page, it is given the wrong one.  I am not sure how the system decides how to ask for a page.

I can't see 1 working as it is exactly the same page, it is just I invoked the "copy para to clipboard" from another instance.  So it would have the same ID?

When a command is invoked, how does it know what page it was invoked from and how does it know which instance (which active window) it was invoked from?

**Comment by stevencohn** [comment]
Yes I can reproduce it

**Comment by jasonjac2** [comment]
This has bitten me a few times in the last week..... I have been reviewing a large project and putting in links on various pages to various other pages to make it easier for a novice to consume.  Having a bunch of pages open and editing, putting links all over the place. I realised that on another virtual desktop I had an instance on ON open and on the main page I was referencing. The outcome was a bunch of "random" links to the main page throughout the notes...... Just enough to convince a newbie I was off my rocker and trying to confuse them more :-).

Also, came across another expression of the project: #1585

Cheers for all the hard work.... still building up to setting the environment to build this.

**Comment by jasonjac2** [comment]
Bitten again - Doing documentation for a project, one page with bits of development, project mgmt.  Pulling together references for a meeting.  All on muscle memory. Reviewed my list on links and one liners, only to find all the reference links pointed to the same para on the source page :-(. 

Dev: Did refresh my fork and opened solution, got the first set of requirements, but wouldn't compile - as expected as I hadn't completed the pre-req's for compile..... but then ran out of time.  As an aside where should i be looking for the "get the XML for the current page with this name", which is where I assume the issue lies.

**Comment by stevencohn** [comment]
Hi @jasonjac2. I believe this to be an internal problem with the OneNote API. After opening up two windows viewing the same page, the API thinks the context is the first window. It's not until you modify the page content in a secondary window will OneNote set the context of the API to that window. You can see replicate this just using the Show XML command as you suggested.

**Comment by jasonjac2** [comment]
Hi @stevencohn I can't replicate this, i.e. 
- 2 instances on same page, caret at different points in page.
- 3rd instance on a new page
- Reproduce the issue, 2nd instance right click | copy link to para
- goto 3rd instance and paste - link is to caret on 1st instance.
- Edit 2nd instance -I added a space 1st time and a couple of carriage returns the second time
- redo test, you still get the link to the 1st instance.
Version 6.5.1 with OneNote 16.0.18025.20160

**Comment by jasonjac2** [comment]
@stevencohn so on this one, it looks like for me the editing of the page "flipping" ON into realising which the current Instance of a page doesn't actually make any difference.  For me, it is predictable which instance it will use, based on (I think) the either the ordering of them being opened OR the order of them on the taskbar (when I press the ON icon on the taskbar and it shows me n instances).
I know MS are not very responsive, but do you know if it s a known bug?  If not can you / we raise a bug report?

**Comment by stevencohn** [comment]
Yes, that's possible

**Comment by jasonjac2** [comment]
I would count that as a weak bite.  Kind of like my answer to questions in development meetings "anything is possible". :-)

**Comment by stevencohn** [comment]
Ha! Yes it's quite possible and easy enough to add to the next release. Better?

**Comment by stevencohn** [comment]
I tried to enter a question myself but the Microsoft Q&A site is crap. It accepted [my question](https://learn.microsoft.com/en-us/answers/questions/2122435/onenote-interop-api-targets-wrong-window-instance), and I can see it in the "My content" list, but clicking it brings me to a 404 Page Not Found 😦 Maybe it's a timing thing. Anyway, here's what I wrote:

_C# with .NET Framework 4.8, Microsoft® OneNote® 2021 MSO (Version 2410 Build 16.0.18129.20100) 64-bit_ 

Open up two OneNote windows viewing the same page. From the second window, invoke an add-in command that uses the managed Interop API.

Any queries or updates made by the API will target the first OneNote window instance, even while the second window instance is active and has focus.

However, if the content of the page is manually modified from the second [active] window instance, then and only then will the API call target and update the page in the second window.

Expectation is that the update content will be visible immediately in the active OneNote window instance.

**Comment by jasonjac2** [comment]
I didn't even know https://learn.microsoft.com/ was a thing!

**Comment by stevencohn** [comment]
Closing this since there nothing further I can do at this time.

**Comment by stevencohn** [comment]
There is not perfect solution. This is indeed a COM Interop API problem. Various workarounds were attempted - all failed.

The only "solution" is to determine if this current page is open in at least one other OneNote window and warn the user that the copy link to paragraph may fail and ask if they want to continue. If they know they're on the "primary" window (the first one opened) then they can choose to continue and it should work fine. But secondary windows will always copy a link to the paragraph marked as current in the primary window.

**Comment by jasonjac2** [comment]
exactly, it makes you be extra observant about what you paste (e.g. you'll click it) OR you'll realise you have opened multiple instances of that page and close all but one (which is what I would do 90% of the time).
If you could do it with a clippy animation "I think there is a 50/50 chance of making a mistake" with an audible ding ding :-).

<img width="100" height="100" alt="Image" src="https://github.com/user-attachments/assets/653ef7c0-af67-4188-8453-a8daabf19606" />
Image courtesy of: https://blog.vjeux.com/2024/project/clippy-gifs.html

**Comment by jasonjac2** [comment]
Tested in beta 7.1.1 B3 - works beautifully.
